### PR TITLE
feat: :sparkles: proper tail recursion

### PR DESCRIPTION
**Pull request by CurryPseudo**

### CODE REVIEW

File: src/interpreter/mod.rs [hunk 1]
```diff
@@ -284,39 +284,64 @@ impl<R: RealNumberInternalTrait, E: IEnvironment<R>> Interpreter<R, E> {
...
 
+    fn apply_scheme_procedure(
+        formals: &Vec<String>,
+        internal_definitions: &Vec<Definition>,
+        expressions: &Vec<Expression>,
+        closure: Rc<RefCell<E>>,
+        args: ArgVec<R, E>,
+        is_tail: bool,
+    ) -> Result<Value<R, E>> {
+        let mut _local_env = None;
```
Comment (Danielmelody) [inline src/interpreter/mod.rs:320]: variable should not begin with an underscore.
https://doc.rust-lang.org/1.0.0/style/style/naming/README.html

**Review by Danielmelody — CHANGES_REQUESTED** [review]
You may need to test examples/fib.scm to actually benchmark the tail recurson optimization.

File: src/interpreter/mod.rs (hunk 1 above)
Comment (Danielmelody) [reply src/interpreter/mod.rs:320]: And also, try add some tests?

Comment (CurryPseudo) [reply src/interpreter/mod.rs:320]: > 
> 
> variable should not begin with an underscore.
> https://doc.rust-lang.org/1.0.0/style/style/naming/README.html


```rust
        let mut local_env = None;
        let closure_ref = if is_tail {
            closure.borrow_mut().clear_definition();
            &closure
        } else {
            local_env = Some(Rc::new(RefCell::new(E::new_child(closure.clone()))));
            &local_env.as_ref().unwrap()
        };
```
If I write the code like this, rustc will warn:
warning: value assigned to `local_env` is never read
   --> src\interpreter\mod.rs:320:13
    |
320 |         let mut local_env = None;
    |             ^^^^^^^^^^^^^
    |
    = note: `#[warn(unused_assignments)]` on by default
    = help: maybe it is overwritten before being read?

warning: 1 warning emitted

Comment (CurryPseudo) [reply src/interpreter/mod.rs:320]: > 
> 
> And also, try add some tests?

Yeah, I do add a long test in parser::definition([parser.rs](https://github.com/Danielmelody/Ruschm/blob/feat-proper-tail-recursion/src/parser.rs#L732-L799)), and edit some old test to suit the change([parser.rs](https://github.com/Danielmelody/Ruschm/blob/feat-proper-tail-recursion/src/parser.rs#L720), [parser.rs](https://github.com/Danielmelody/Ruschm/blob/feat-proper-tail-recursion/src/parser.rs#L858), [parser.rs](https://github.com/Danielmelody/Ruschm/blob/feat-proper-tail-recursion/src/parser.rs#L896))

**Pull request closed without merging**